hack/verify-license.py

```diff
@@ -0,0 +1,945 @@
+    'db170a1a3711f86e4ce228a9fe74e101315900f8':
+        ('vendor/github.com/JeffAshton/win_pdh/LICENSE', 'BSD (3-Clause)'),
+}
+
+# Ignore replacements in these forms.
+# Format is from_string: {target_string}
+IGNORE_SUBS = {
+    '': {
+        'The MIT License (MIT)', 'The MIT License', 'MIT License',
+        'ISC License',
+        '>', '> >'
+    },
+    '1.': {'', '*', '-'},
+    '2.': {'', '*', '-'},
+    '3.': {'', '*'},
+    '(a)': {''}, '(b)': {''}, '(c)': {''}, '(d)': {''},
+    'HOLDER': {'OWNER'},
+    'OWNER': {'HOLDER'},
+    'http://www.apache.org/licenses/': {'https://www.apache.org/licenses/'},
+    'http://www.apache.org/licenses/LICENSE-2.0': {
+        'https://www.apache.org/licenses/LICENSE-2.0'},
+    ('The views and conclusions contained in the software and documentation are'
+     ' those of the authors and should not be interpreted as representing '
+     'official policies, either expressed or implied, of the FreeBSD Project.'):
+     {''}
+}
+
+# These strings are what *should* be replaced in licenses to indicate years and owners.
+COPYRIGHT_NAMES = {
+    '<YEAR> <COPYRIGHT HOLDER>',
+    '<YEAR> <OWNER>',
+    '[yyyy] [name of copyright owner]',
+    'copyright holder',
+    'the copyright holder',
+}
+
+# Ignore punctuation differences.
+PUNCT_RE = re.compile(r'[][(){},.]+')
+
+
+def find_licenses(root):
+    '''
+    Find all license files by walking from a root.
+
+    Args:
+        root: the directory to search for licenses inside
+    Returns:
+        licenses: a list of LICENSE files to validate.
+        uncovered: a list of directories that are missing LICENSE files.
+    '''
+    licenses = []
+    uncovered = []
+    for dirpath, dirnames, filenames in os.walk(root, followlinks=True):
+        if dirpath in WEIRD_LICENSE_NAMES:
+            path = os.path.join(dirpath, WEIRD_LICENSE_NAMES[dirpath])
+            if os.path.exists(path):
+                licenses.append(path)
+                dirnames[:] = []  # don't walk further
+                continue
+        for filename in filenames:
+            base, ext = os.path.splitext(filename)
+            if base.lower() in ('licence', 'license') \
+                    and ext in ('', '.txt', '.md', '.code'):
+                licenses.append(os.path.join(dirpath, filename))
+                dirnames[:] = []  # don't walk further
+                break
+        else:  # no license found
+            for filename in filenames:
+                if filename in ('BUILD', 'OWNERS'):
+                    continue  # ignore BUILD files
+                fpath = os.path.join(dirpath, filename)
+                if any(fnmatch.fnmatch(fpath, pat) for pat in IGNORE_PATTERNS):
+                    continue
+                uncovered.append(dirpath)
+                dirnames[:] = []  # don't walk further
+                break
+    licenses.sort()
+    uncovered.sort()
+    return licenses, uncovered
+
+
+def extract_substitutions(diff):
```

Comment:
Please include unit tests for this module

Reply:
done (doctests)